kernel/cbytegen.ml

```diff
@@ -593,102 +449,104 @@ let rec get_alias env kn =
 	| BCalias kn' -> get_alias env kn'
 	| _ -> kn)
 
+(* spiwack: additional function which allow different part of compilation of the
+      31-bit integers *)
+
+let make_areconst n else_lbl cont =
+  if n <= 0 then
+    cont
+  else
+    Kareconst (n, else_lbl)::cont
+
 (* sz is the size of the local stack *)
-let rec compile_constr reloc c sz cont =
+let rec compile_lam env reloc lam sz cont =
   set_max_stack_size sz;
-  match kind c with
-  | Meta _ -> invalid_arg "Cbytegen.compile_constr : Meta"
-  | Evar _ -> invalid_arg "Cbytegen.compile_constr : Evar"
-  | Proj (p,c) ->
-     let kn = Projection.constant p in
-     let cb = lookup_constant kn !global_env in
-     let pb = Option.get cb.const_proj in
-     let n = pb.proj_arg in
-     compile_constr reloc c sz (Kproj (n,kn) :: cont)
-
-  | Cast(c,_,_) -> compile_constr reloc c sz cont
-
-  | Rel i -> pos_rel i reloc sz :: cont
-  | Var id -> pos_named id reloc :: cont
-  | Const (kn,u) -> compile_const reloc kn u [||] sz cont
-  | Ind (ind,u) ->
-     let bcst = Bstrconst (Const_ind ind) in
+  match lam with
+  | Lrel(_, i) -> pos_rel i reloc sz :: cont
+
+  | Lval v -> compile_structured_constant reloc v sz cont
+
+  | Lproj (n,kn,arg) ->
+     compile_lam env reloc arg sz (Kproj (n,kn) :: cont)
+
+  | Lvar id -> pos_named id reloc :: cont
+
+  | Lconst (kn,u) -> compile_constant env reloc kn u [||] sz cont
+
+  | Lind (ind,u) ->
     if Univ.Instance.is_empty u then
-      compile_str_cst reloc bcst sz cont
-    else
-      comp_app compile_str_cst compile_universe reloc
-	   bcst
-	   (Univ.Instance.to_array u)
-	   sz
-	   cont
-  | Sort (Sorts.Prop _) | Construct _ ->
-      compile_str_cst reloc (str_const c) sz cont
-  | Sort (Sorts.Type u) ->
+      compile_structured_constant reloc (Const_ind ind) sz cont
+    else comp_app compile_structured_constant compile_universe reloc
+        (Const_ind ind) (Univ.Instance.to_array u) sz cont
+
+  | Lsort (Sorts.Prop _ as s) ->
+    compile_structured_constant reloc (Const_sorts s) sz cont
+  | Lsort (Sorts.Type u) ->
      (* We separate global and local universes in [u]. The former will be part
         of the structured constant, while the later (if any) will be applied as
         arguments. *)
      let open Univ in begin
       let u,s = Universe.compact u in
       (* We assume that [Universe.type0m] is a neutral element for [Universe.sup] *)
-      if List.is_empty s then
-        compile_str_cst reloc (Bstrconst (Const_sorts (Sorts.Type u))) sz cont
+      if List.is_empty s then Kconst (Const_sorts (Sorts.Type u)) :: cont
```

Comment:
Since you have introduced `compile_structured_constant`, it is a bit strange not to use it here.